src/Compilers/Core/Portable/VersionHelper.cs

```diff
@@ -74,35 +74,50 @@ private static bool TryParse(string s, bool allowWildcard, ushort maxValue, out
 
             if (hasWildcard)
             {
-                int seconds = ((int)(DateTime.Now.TimeOfDay.TotalSeconds)) / 2;
-                if (seconds > (int)maxValue)
+                for (int i = lastExplicitValue; i < values.Length; i++)
                 {
-                    version = null;
-                    return false;
-                }
-                values[3] = (ushort)seconds;
-
-                if (elements.Length == 3)
-                {
-                    TimeSpan days = DateTime.Today - new DateTime(2000, 1, 1);
-                    int build = (int)days.TotalDays;
-
-                    if (build < 0 || build > (int)maxValue)
-                    {
-                        //alink would generate an error here saying "Cannot auto-generate build and 
-                        //revision version numbers for dates previous to January 1, 2000." Without
-                        //some refactoring here to relay the date problem, Roslyn
-                        //will generate an inaccurate error about the version string being of the wrong format.
-
-                        version = null;
-                        return false;
-                    }
-                    values[2] = (ushort)build;
+                    values[i] = ushort.MaxValue;
                 }
             }
 
             version = new Version(values[0], values[1], values[2], values[3]);
             return true;
         }
+
+        /// <summary>
+        /// If build and/or revision numbers are 65535 they are replaced with time-based values.
+        /// </summary>
+        public static Version GenerateVersionFromPatternAndCurrentTime(DateTime time, Version pattern)
+        {
+            if (pattern == null || pattern.Revision != ushort.MaxValue)
+            {
+                return pattern;
+            }
+
+            // MSDN doc on the attribute: 
+            // "The default build number increments daily. The default revision number is the number of seconds since midnight local time 
+            // (without taking into account time zone adjustments for daylight saving time), divided by 2."
+            if (time == default(DateTime))
+            {
+                time = DateTime.Now;
+            }
+
+            int revision = (int)time.TimeOfDay.TotalSeconds / 2;
+
+            // 24 * 60 * 60 / 2 = 43200 < 65534
+            Debug.Assert(revision < 0xffff);
+
+            if (pattern.Build == ushort.MaxValue)
+            {
+                TimeSpan days = time.Date - new DateTime(2000, 1, 1);
+                int build = Math.Min(ushort.MaxValue, (int)days.TotalDays);
```

Comment:
It is not invalid. It's only considered invalid in the argument of AssemblyVersionAttribute. It's an artificial limitation imposed by the compiler (presumably native compilers did that to be able to represent *). Metadata allows version numbers [0, 0xffff].